core/iwasm/libraries/lib-wasi-threads/test/sync_primitives.h

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (C) 2023 Amazon.com Inc. or its affiliates. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ */
+
+#include <stdbool.h>
+
+/* Mutex */
+
+typedef int mutex_t;
+
+void
+mutex_init(mutex_t *mutex)
+{
+    *mutex = 0;
+}
+
+static bool
+try_mutex_lock(mutex_t *mutex)
+{
+    int expected = 0;
+    return __atomic_compare_exchange_n(mutex, &expected, 1, false,
+                                       __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+}
+
+void
+mutex_lock(mutex_t *mutex)
+{
+    while (!try_mutex_lock(mutex))
+        __builtin_wasm_memory_atomic_wait32(mutex, 1, -1);
+}
+
+void
+mutex_unlock(mutex_t *mutex)
+{
+    __atomic_store_n(mutex, 0, __ATOMIC_SEQ_CST);
+    __builtin_wasm_memory_atomic_notify(mutex, 1);
+}
+
+/* Barrier */
+
+typedef struct {
+    int count;
+    int num_threads;
+    int mutex;
+    int ready;
+} barrier_t;
+
+void
+barrier_init(barrier_t *barrier, int num_threads)
+{
+    barrier->count = 0;
+    barrier->num_threads = num_threads;
+    barrier->mutex = 0;
+    barrier->ready = 0;
+}
+
+void
+barrier_wait(barrier_t *barrier)
+{
+    bool no_wait;
+
+    mutex_lock(&barrier->mutex);
+    barrier->count++;
+    no_wait = (barrier->count >= barrier->num_threads);
+    mutex_unlock(&barrier->mutex);
+
+    if (no_wait) {
+        __atomic_store_n(&barrier->ready, 1, __ATOMIC_SEQ_CST);
+        __builtin_wasm_memory_atomic_notify(&barrier->ready, 1);
```

Comment:
Why do you notify only 1 thread?

Reply:
Okay, it looked like it works with notifying only 1 thread because in such sample it works wrong on main_proc_exit_wait:
-> main thread starts actually waiting
-> 2 threads enter atomic.wait, but not actually waiting yet
->  3 thread increases the counter and notifies main thread
-> main thread wakes up and calls proc_exit before any other thread actually started waiting

I don't think it's possible to make a barrier without a cond_var

Reply:
> Why do you notify only 1 thread?

Nice catch, thanks! It got lost during the copy and paste.

> Okay, it looked like it works with notifying only 1 thread because in such sample it works wrong on main_proc_exit_wait:

No, I think it should be fine, it's an `atomic.wait` not a `pthread_cond_wait`, the condition variable is part of its implementation (one of the argument is `expect` in fact).

Reply:
Okay, you fixed number of notifying threads. But there is a possibility for deadlock with this logic.
- all threads leave section under lock
- counter has reached number of threads, so no_wait set to true in 1 of the threads
- this thread notifies all the other threads before the threads reached waiting call
-> all other threads hang infinetely

Reply:
Managed to hang it without changes from #2024, but with this changes it doesn't seem to be a problem